### PR TITLE
Add process/pipes IPC test

### PR DESCRIPTION
**Pull request by stephentoub**
Add a test for inter-process communication using pipes.

### CODE REVIEW

**Comment by stephentoub** [comment]
cc: @pallavit, @krwq

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs [hunk 1]
```diff
@@ -161,7 +163,7 @@ public void Process_ExitCode()
                 Process p = CreateProcess();
                 p.Start();
                 Assert.True(p.WaitForExit(WaitInMS));
-                Assert.Equal(p.ExitCode, 100);
+                Assert.Equal(p.ExitCode, SuccessExitCode);
```
Comment (Maxwe11) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Nit: expected `SuccessExitCode` should be passed as first argument

**Comment by mellinoe** [comment]
The test looks good to me, should we put it in with the pipes tests, though?

Also, we have a few more pipes tests in TFS that do some inter-process stuff. At the time I put out the tests we didn't have a standardized way of launching multiple test processes (I'm not sure we were even running on CoreCLR at the time actually..). Eventually we should take a look at those and see if we should bring those out.

**Comment by stephentoub** [comment]
> The test looks good to me

Thanks.

> should we put it in with the pipes tests, though?

I debated that.  It's kind of a pipes thing, and kind of a process thing, and since we already have support in the process tests for launching processes, cleaning them up, etc., and since we already have a target process exe, seemed to make sense to put it there. Let me know if you feel strongly about moving it. If we end up adding a larger suite of these IPC tests, I agree we should find another place for them and move this then.

> Eventually we should take a look at those and see if we should bring those out.

Can you open an issue for that?

File: src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/System.Diagnostics.Process.Tests/ProcessTest.cs]: Fixed.

**Comment by pallavit** [comment]
>  debated that. It's kind of a pipes thing, and kind of a process thing, and since we already have support in the process tests for launching processes, cleaning them up, etc., and since we already have a target process exe, seemed to make sense to put it there.

In my opinion they are more like integration testing than component testing. Do we have a plan for common tests?

**Comment by mmitche** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
> In my opinion they are more like integration testing than component testing. 

Sort of.  But the primary use case for pipes is cross-process communication, so it really begs for a unit test around IPC.  And an important aspect of Process launching is whether handles are inherited and the like, so again it begs for a unit test to exercise that.  This is one unit test then that happens to test both pieces.  In that sense it's an integration test, but regardless of what we call it, we should have a test like this that runs as part of our inner loop testing.

> Do we have a plan for common tests?

Don't know. @weshaggard or @joshfree can probably speak to plans beyond unit tests best.

**Comment by pallavit** [comment]
I agree with you that these scenarios are important and should be tested. Thanks for closing the gaps here. However, I would expect more such situations in future and so was curious to know our plans around multiple components testing

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by weshaggard** [comment]
I wouldn't say we have a plan exactly. I don't mind use the TestConsoleApp for multiple purposes but I don't like doing Pipes specific library testing in the Process test project. I would much prefer those test be included as part of the Pipes test project. As Pipes are using Process as support for testing it's APIs and not the other way around so the dependency seems reversed to me.

**Comment by stephentoub** [comment]
> I would much prefer those test be included as part of the Pipes test project

@weshaggard, if you feel strongly about it, please open an issue to track.  I'd be ok with us moving this, in particular if we were going to add a bunch more, but keep in mind that it took weeks to stabilize the process tests with regards to the CI infrastructure, cleaning up after zombie processes, and the like, and we'll likely face some of that again if we introduce yet additional test projects that spawn, manipulate, and synchronize with helper processes, so this isn't just a matter of trivial copy-and-paste.  In the meantime, from a practical perspective I personally don't see anything wrong with this one test living here, in particular (as I noted in a previous response in this thread) because it helps to verify something that is Process-specific, that handles can be inherited... if we didn't have this test, we'd need to introduce another one that opened some form of inheritable handle and verified it could be used in another process... with this we get two for the price of one.  My $.02.

**Comment by weshaggard** [comment]
@stephentoub I don't have any major concern keeping this test here but in general I want us to try and minimize the dependencies our tests are taking. My general goal is to try and have the same exact dependencies between the product library and the test library testing that product library. That way we can use those same tests to validate that library on any platform that it is supported on. The more dependencies we take the more risk we have in making the tests not runnable in other scenarios. Again that is a general rule and rules can always be broken for the right reasons, and for this one case maybe the reliability is a good enough reason to keep it here. However if we end up feeling like we want to grow this set we should reconsider factoring them out.

**Comment by stephentoub** [comment]
Thanks, Wes. I think we're generally agreed.

**Pull request merged**